### PR TITLE
qemu: update to 6.1.0, add usb passthrough option

### PR DESCRIPTION
**Pull request by vooon**
Signed-off-by: Vladimir Ermakov <vooon341@gmail.com>

Maintainer: @yousong
Compile tested: x86_64, OpenWrt master (https://github.com/openwrt/openwrt/commit/90e167abaae80a59c8780d3305f8e1260d9f1339), GCC 10.3
Run tested: x86_64 host, KVM for qemu-x86_64-softmmu + UEFI VM (homeassistant ova).

Description:
- Version updated to 6.1.0
- I had have compile problems with gcc 10 on 5.x
- Added option to enable USB pass-through (used it before with 5.x)
- Some changes in build system do not allow to use `make x86_64-softmmu` or `make qemu-img`, current workaround - build all

### CODE REVIEW

**Comment by vooon** [comment]
Works perfectly fine for my home assistant ova.
```
# cat /etc/init.d/qemu-homeassistant 
#!/bin/sh /etc/rc.common

#set -x

START=99
STOP=1
 
qemu_pidfile="/var/run/qemu-homeassistant.pid"
qemu_logfile="/srv/log/qemu-homeassistant.log"
qemu_uuid="88cad999-3483-48cb-b324-ea62dc641ed9"
qemu_qmp_port=4445

start() {
qemu-system-x86_64 -enable-kvm -cpu host -smp 2 -m 2G \
	-uuid $qemu_uuid \
	-drive if=pflash,file=/srv/vm/OVMF/OVMF_CODE.fd,format=raw,readonly=on \
	-drive if=pflash,file=/srv/vm/OVMF/homeassistant_vars.fd,format=raw \
	-drive file=/srv/vm/homeassistant.qcow2,if=virtio \
	-drive file=/mnt/share/vm/ha-media.qcow2,if=virtio \
	-device virtio-net-pci,mac=E2:F2:6A:01:9D:D0,netdev=br0 \
	-netdev bridge,br=br-lan,id=br0 \
	-serial pty \
	-serial pty \
	-device qemu-xhci \
	-device usb-host,vendorid=0x0a12,productid=0x0001 \
	-vnc :0 \
	-chardev socket,path=/tmp/ha-qga.sock,server,nowait,id=qga0 \
	-device virtio-serial \
	-device virtserialport,chardev=qga0,name=org.qemu.guest_agent.0 \
	-qmp tcp:127.0.0.1:$qemu_qmp_port,server,nowait \
	-daemonize &> $qemu_logfile

	#-serial /dev/tty-senseair-s8-co2 \
	#-serial /dev/tty-zzh \
	#-device usb-host,vendorid=0x0a12,productid=0x0001 \

#/usr/bin/pgrep qemu-system-x86_64 > $qemu_pidfile
/usr/bin/pgrep -f "qemu-system-x86_64.*$qemu_uuid" > $qemu_pidfile
echo "QEMU: Started VM with PID $(cat $qemu_pidfile)."
}
 
stop() {
echo "QEMU: Sending 'system_powerdown' to VM with PID $(cat $qemu_pidfile)."
nc localhost $qemu_qmp_port <<QMP 
{ "execute": "qmp_capabilities" } 
{ "execute": "system_powerdown" } 
QMP
 
if [ -e $qemu_pidfile ]; then
	if [ -e /proc/$(cat $qemu_pidfile) ]; then
		echo "QEMU: Waiting for VM shutdown."
		while [ -e /proc/$(cat $qemu_pidfile) ]; do sleep 1s; done
		echo "QEMU: VM Process $(cat $qemu_pidfile) finished."
	else
		echo "QEMU: Error: No VM with PID $(cat $qemu_pidfile) running."
	fi
 
	rm -f $qemu_pidfile
else
	echo "QEMU: Error: $qemu_pidfile doesn't exist."
fi
}
```

**Comment by neheb** [comment]
Need to run make package/qemu/refresh

**Comment by vooon** [comment]
@neheb done.

**Comment by vooon** [comment]
@neheb do you know what flag is needed to have python3 setuptools on host?
```
../meson.build:56:0: ERROR: <ExternalProgram 'python3' -> ['/home/build/openwrt/staging_dir/host/bin/python3']> is not a valid python or it is missing setuptools
```

**Comment by vooon** [comment]
That probably bad idea to add python3/host to build depends, but i don't know much about meson and qemu's use of it.
And since qemu does a lot of code generation it is likely that python was being imported in many build scripts.

**Comment by neheb** [comment]
Yes I do. See https://github.com/openwrt/packages/blob/master/multimedia/gstreamer1/patches/020-distutils.patch

**Comment by vooon** [comment]
I patched meson.build, let's see would it work. For previous one it's not clear why x86 fail.

**Comment by neheb** [comment]
Need to refresh patches.

**Comment by vooon** [comment]
Ugh, no, i cannot make it build with patch 0008. reverted back to python3/host.

**Comment by neheb** [comment]
ping @jefferyto @dhewg 

Not sure about the python3 host dependency...

**Comment by jefferyto** [comment]
@vooon perhaps you can post the build error you get after patching meson.build

**Comment by vooon** [comment]
@jefferyto , @neheb i wanted to reproduce build error to post here, but instead i got it build.
But I updated openwrt and rebased packages repos. To re-check i'm started clean rebuild again.

**Comment by vooon** [comment]
@neheb friendly ping.

**Comment by neheb** [comment]
python3/host is no go. It increases build time too much.

**Comment by neheb** [comment]
While this is converted to meson, I still see old Makefile type patches. Intentional?

**Comment by neheb** [comment]
Doesn't compile on mips locally

```
../tests/fp/fp-bench.c: In function 'set_host_precision':
../tests/fp/fp-bench.c:491:17: error: 'FE_TOWARDZERO' undeclared (first use in this function)
  491 |         rhost = FE_TOWARDZERO;
      |                 ^~~~~~~~~~~~~
../tests/fp/fp-bench.c:491:17: note: each undeclared identifier is reported only once for each function it appears in
../tests/fp/fp-bench.c:494:17: error: 'FE_DOWNWARD' undeclared (first use in this function)
  494 |         rhost = FE_DOWNWARD;
      |                 ^~~~~~~~~~~
../tests/fp/fp-bench.c:497:17: error: 'FE_UPWARD' undeclared (first use in this function)
  497 |         rhost = FE_UPWARD;

```

Interesting that tests are being run...

**Comment by neheb** [comment]
Another one...

```
../block/export/fuse.c: In function 'fuse_fallocate':
../block/export/fuse.c:563:23: error: 'FALLOC_FL_ZERO_RANGE' undeclared (first use in this function)
  563 |     } else if (mode & FALLOC_FL_ZERO_RANGE) {
      |                       ^~~~~~~~~~~~~~~~~~~~
../block/export/fuse.c:563:23: note: each undeclared identifier is reported only once for each function it appears in
../block/export/fuse.c: In function 'fuse_lseek':
../block/export/fuse.c:639:19: error: 'SEEK_HOLE' undeclared (first use in this function)
  639 |     if (whence != SEEK_HOLE && whence != SEEK_DATA) {
      |                   ^~~~~~~~~
../block/export/fuse.c:639:42: error: 'SEEK_DATA' undeclared (first use in this function); did you mean 'SEEK_SET'?
  639 |     if (whence != SEEK_HOLE && whence != SEEK_DATA) {
      |                                          ^~~~~~~~~
      |                                          SEEK_SET
ninja: build stopped: subcommand failed.
```

**Comment by vooon** [comment]
> While this is converted to meson, I still see old Makefile type patches. Intentional?

I planned to rebase them out after CI pass.

> Doesn't compile on mips locally

I didn't try to build there, only amd64 and armv7.
Could you please provide config file?

**Comment by vooon** [comment]
@neheb note, i updated PR to 6.1.0, which i'm testing now.

File: utils/qemu/Makefile
Comment (neheb) [inline utils/qemu/Makefile:444]: Sounds like this should be removed.

File: utils/qemu/Makefile
Comment (neheb) [inline utils/qemu/Makefile:451]: Sounds like this should be removed.

File: utils/qemu/Makefile
Comment (neheb) [inline utils/qemu/Makefile:13]: Could replace with $(AUTORELEASE)

File: utils/qemu/Makefile
Comment (neheb) [inline utils/qemu/Makefile:17]: This should be either -only or -or-later, as per SPDX.

File: utils/qemu/Makefile
Comment (neheb) [inline utils/qemu/Makefile:276]: Interesting that CONFIGURE_ARGS is used for a meson package...

File: utils/qemu/patches/0008-meson-python-fix.patch
Comment (neheb) [inline utils/qemu/patches/0008-meson-python-fix.patch:1]: This patch is no longer needed.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile:17]: I'm not quite sure what should be used: https://github.com/qemu/qemu/blob/master/LICENSE

It sounds to me, that most of the code is GPL-2.0-only, but some parts could be GPL-2.0-or-later.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile:450]: Will do.

File: utils/qemu/patches/0008-meson-python-fix.patch
Comment (vooon) [reply utils/qemu/patches/0008-meson-python-fix.patch:1]: Yes, i removed that patch and it is building. Note i'm not yet got anything working. Seems at lease x86 need libbpf now.

File: utils/qemu/Makefile
Comment (neheb) [reply utils/qemu/Makefile:17]: Use -only then.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile]: Done.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile]: Done.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile]: Done.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile:276]: It still uses `./configure`. Do you think that it should be changed?

File: utils/qemu/patches/0008-meson-python-fix.patch
Comment (vooon) [reply utils/qemu/patches/0008-meson-python-fix.patch]: Qemu compilled. Will update my router soon...

File: utils/qemu/Makefile
Comment (neheb) [reply utils/qemu/Makefile:443]: I mean the entire Build/Compile section

**Comment by vooon** [comment]
@neheb tested on x86_64 - all my VMs are working. USB and serial pass-trough works.

**Review by neheb — APPROVED** [review]

**Comment by neheb** [comment]
ping @yousong

File: utils/qemu/Makefile
Comment (yousong) [inline utils/qemu/Makefile:301]: Please add a note about why this was removed

**Review by yousong — APPROVED** [review]
Maybe we should squash the two commits together for final merge.  I am sure this version bump requires many diggings and tryings to have such shape.  Please considering adding your thoughts and findings in the commit message for other devs to learn the details.  It will be useful for future maintenance.

**Comment by vooon** [comment]
@yousong done rebase. Added some notes to the commit.
But actually there are not much findings. Perhaps primary one - how to fix meson to use correct python.
Other changes mostly result of build trials...

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile:301]: That option was removed from qemu:
```
qemu-6.1.0$ ./configure --help | grep vhost
  vhost-net       vhost-net kernel acceleration support
  vhost-vsock     virtio sockets device support
  vhost-scsi      vhost-scsi kernel target support
  vhost-crypto    vhost-user-crypto backend support
  vhost-kernel    vhost kernel backend support
  vhost-user      vhost-user backend support
  vhost-user-blk-server    vhost-user-blk server support
  vhost-vdpa      vhost-vdpa kernel backend support
```

Comment (yousong) [reply utils/qemu/Makefile:301]: It's still there, just not in the show_help output：https://github.com/qemu/qemu/blob/f9baca549e44791be0dd98de15add3d8452a8af0/configure#L1283-L1286

Maybe try again with the commands in the Makefile comment:

https://github.com/openwrt/packages/blob/f2280ae82e3ec04649639b020e4676f1d2125d13/utils/qemu/Makefile#L355-L363

**Comment by vooon** [comment]
Frinedly ping @neheb .

**Comment by neheb** [comment]
@vooon the meson and python thing was fixed. distutils is a requirement now.

As far as merging, up to @yousong

**Comment by yousong** [comment]
Hi, @vooon , the conversation about --enable-vhost-user-fs is still ongoing ;)

**Comment by vooon** [comment]
@yousong ah, right. I do remember, that i had to remove some of args cause configure failed. Checking.

File: utils/qemu/Makefile
Comment (vooon) [reply utils/qemu/Makefile:301]:
```
user@6b03bfc66c94:~/ws$ qemu_configure=build_dir/target-x86_64_musl/qemu-6.1.0/configure 
user@6b03bfc66c94:~/ws$ openwrt_makefile=feeds/packages/utils/qemu/Makefile 
user@6b03bfc66c94:~/ws$ for arg in $(grep -E '^\s*--disable-[^)]+\)' "$qemu_configure" | cut -f1 -d')'); do grep -qE "(--enable|--disable)${arg#--disable}" "$openwrt_makefile" || echo "$arg"; done
--disable-virtiofsd
--disable-vnc
--disable-gettext
--disable-block-drv-whitelist-in-tools
--disable-tsan
--disable-vnc-sasl
--disable-vnc-jpeg
--disable-vnc-png
--disable-slirp
--disable-nvmm
--disable-malloc-trim
--disable-spice
--disable-spice-protocol
--disable-pie
--disable-safe-stack
--disable-cfi
--disable-cfi-debug
--disable-fdt
--disable-bpf
--disable-blobs
--disable-vhost-user-fs
--disable-u2f
--disable-libusb
--disable-zlib-test
--disable-avx2
--disable-avx512f
--disable-virtio-blk-data-plane|--enable-virtio-blk-data-plane
--disable-git-update
--disable-fuzzing
--disable-rng-none
--disable-keyring
--disable-libdaxctl
--disable-fuse
--disable-fuse-lseek
--disable-multiprocess
--disable-gio
--disable-slirp-smbd
```

Comment (vooon) [reply utils/qemu/Makefile:301]: Checked with vhost-user-fs - all ok.

**Comment by vooon** [comment]
@yousong reverted back vhost-user-fs - build successful. What do you suppose to do with other options?

**Comment by yousong** [comment]
> @yousong reverted back vhost-user-fs - build successful. What do you suppose to do with other options?

I do not have much experience with those other options, e.g. `--disable-sheepdog`, `--disable-vxhs`.   As such and as current maintainer I prefer there is a note in the commit message sharing with us why the changes were made.

**Comment by vooon** [comment]
@yousong added a note to the commit message. Now i recall why i got unknown options passed to configure.

**Comment by yousong** [comment]
Merged.  Thank you all for the time and efforts!

**Comment by neheb** [comment]
https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a53/packages/qemu/compile.txt

**Comment by vooon** [comment]
@neheb seems we need `--disable-fuse-lseek` for that targets, even though it's support detected.
```
                          libudev: NO
                       FUSE lseek: YES
```

**Comment by neheb** [comment]
Nah it's just a missing header.

**Comment by vooon** [comment]
As for x86_64 i do not have that error on local builds.

**Comment by vooon** [comment]
Ah. Ok.

**Pull request merged**
